### ISSUE TITLE
Using break in "for x in xxx" loop causes error

### ISSUE THREAD

**Issue by ZiwKerman** [KSP-KOS/KOS#694]
![terminal](https://cloud.githubusercontent.com/assets/11242352/6906774/bd06e17e-d728-11e4-930c-920f1870de82.PNG)

Script in question:

```javascript
    print "Trying to deploy satellite".

	set deployDocks to ship:partstagged("SatDeployDock").

	for d in deployDocks
	{
		set d1 to d:getmodule("ModuleDockingNode").
		if d1:allevents[0] = "(callable) decouple node, is KSPEvent"
		{
			d1:doevent("decouple node").
			print "Decoupling Sat".
			break.
		}
	}
```

Output log:

```
	kOS: Archive: Getting File By Name: sat_deploy
	 
	(Filename: C:/BuildAgent/work/d63dfc6385190b60/artifacts/StandalonePlayerGenerated/UnityEngineDebug.cpp Line: 49)

	kOS: Volume: Add: sat_deploy.ks
	 
	(Filename: C:/BuildAgent/work/d63dfc6385190b60/artifacts/StandalonePlayerGenerated/UnityEngineDebug.cpp Line: 49)

	kOS: At sat_deploy on archive, line 16
	}
	^

	 
	(Filename: C:/BuildAgent/work/d63dfc6385190b60/artifacts/StandalonePlayerGenerated/UnityEngineDebug.cpp Line: 49)

	kOS.Safe.Exceptions.KOSUndefinedIdentifierException: Undefined Variable Name 'deploydocks-iterator'. 

	  at kOS.Execution.CPU.GetVariable (System.String identifier, Boolean barewordOkay, Boolean failOkay) [0x00000] in <filename unknown>:0 

	  at kOS.Execution.CPU.GetValue (System.Object testValue, Boolean barewordOkay) [0x00000] in <filename unknown>:0 

	  at kOS.Execution.CPU.PopValue (Boolean barewordOkay) [0x00000] in <filename unknown>:0 

	  at kOS.Safe.Compilation.OpcodeGetMember.Execute (ICpu cpu) [0x00000] in <filename unknown>:0 

	  at kOS.Execution.CPU.ExecuteInstruction (kOS.Execution.ProgramContext context) [0x00000] in <filename unknown>:0 

	  at kOS.Execution.CPU.ContinueExecution () [0x00000] in <filename unknown>:0 

	  at kOS.Execution.CPU.Update (Double deltaTime) [0x00000] in <filename unknown>:0 
	 
	(Filename: C:/BuildAgent/work/d63dfc6385190b60/artifacts/StandalonePlayerGenerated/UnityEngineDebug.cpp Line: 49)

	Code Fragment

	File                 Line:Col IP   opcode operand

	----                 ----:--- ---- --------------------- 

							0:0   0000 jump 1 

	archive/sat_deploy      3:1   0001 push $<argstart> 

	archive/sat_deploy      3:7   0002 push Trying to deploy satellite 

	archive/sat_deploy      3:7   0003 call print() 

	archive/sat_deploy      3:7   0004 pop 

	archive/sat_deploy      5:5   0005 push $deploydocks 

	archive/sat_deploy      5:20  0006 push $ship 

	archive/sat_deploy      5:25  0007 push partstagged 

	archive/sat_deploy      5:25  0008 getmethod 

	archive/sat_deploy      5:36  0009 push $<argstart> 

	archive/sat_deploy      5:37  0010 push SatDeployDock 

	archive/sat_deploy      5:37  0011 call <indirect> 

	archive/sat_deploy      5:37  0012 store 

	archive/sat_deploy      7:1   0013 pushscope 1 0 

	archive/sat_deploy      7:1   0014 push $deploydocks-iterator 

	archive/sat_deploy      7:10  0015 push $deploydocks 

	archive/sat_deploy      7:10  0016 push iterator 

	archive/sat_deploy      7:10  0017 getmember 

	archive/sat_deploy      7:10  0018 storelocal 

	archive/sat_deploy      7:10  0019 push $deploydocks-iterator 

	archive/sat_deploy      7:10  0020 push next 

	archive/sat_deploy      7:10  0021 getmember 

	archive/sat_deploy      7:10  0022 br.false 40 

	archive/sat_deploy      7:5   0023 push $d 

	archive/sat_deploy      7:5   0024 push $deploydocks-iterator 

	archive/sat_deploy      7:5   0025 push value 

	archive/sat_deploy      7:5   0026 getmember 

	archive/sat_deploy      7:5   0027 storelocal 

	archive/sat_deploy      8:1   0028 pushscope 2 1 

	archive/sat_deploy      9:6   0029 push $d1 

	archive/sat_deploy      9:12  0030 push $d 

	archive/sat_deploy      9:14  0031 push getmodule 

	archive/sat_deploy      9:14  0032 getmethod 

	archive/sat_deploy      9:23  0033 push $<argstart> 

	archive/sat_deploy      9:24  0034 push ModuleDockingNode 

	archive/sat_deploy      9:24  0035 call <indirect> 

	archive/sat_deploy      9:24  0036 store 

	archive/sat_deploy     10:5   0037 push $d1 

	archive/sat_deploy     10:8   0038 push allevents 

	archive/sat_deploy     10:8   0039 getmember 

	archive/sat_deploy     10:18  0040 push 0 

	archive/sat_deploy     10:18  0041 getindex 

	archive/sat_deploy     10:23  0042 push (callable) decouple node, is KSPEvent 

	archive/sat_deploy     10:21  0043 eq 

	archive/sat_deploy     10:21  0044 br.false 16 

	archive/sat_deploy     11:2   0045 pushscope 3 2 

	archive/sat_deploy     12:3   0046 push $d1 

	archive/sat_deploy     12:6   0047 push doevent 

	archive/sat_deploy     12:6   0048 getmethod 

	archive/sat_deploy     12:13  0049 push $<argstart> 

	archive/sat_deploy     12:14  0050 push decouple node 

	archive/sat_deploy     12:14  0051 call <indirect> 

	archive/sat_deploy     12:14  0052 pop 

	archive/sat_deploy     13:3   0053 push $<argstart> 

	archive/sat_deploy     13:9   0054 push Decoupling Sat 

	archive/sat_deploy     13:9   0055 call print() 

	archive/sat_deploy     13:9   0056 pop 

	archive/sat_deploy     14:3   0057 popscope 3 

	archive/sat_deploy     14:3   0058 jump 4 

	archive/sat_deploy     15:2   0059 popscope 1 

	archive/sat_deploy     16:1   0060 popscope 1 

	archive/sat_deploy     16:1   0061 jump -42 

	archive/sat_deploy     16:1   0062 push $deploydocks-iterator 

	archive/sat_deploy     16:1   0063 push reset 

	archive/sat_deploy     16:1   0064 getmember <<--INSTRUCTION POINTER--

	archive/sat_deploy     16:1   0065 pop 

	archive/sat_deploy     16:1   0066 push $deploydocks-iterator 

	archive/sat_deploy     16:1   0067 unset 

	archive/sat_deploy      7:5   0068 push $d 

	archive/sat_deploy      7:5   0069 unset 

	archive/sat_deploy      7:5   0070 popscope 1 

							0:0   0071 EOP 


	 
	(Filename: C:/BuildAgent/work/d63dfc6385190b60/artifacts/StandalonePlayerGenerated/UnityEngineDebug.cpp Line: 49)

	kOS: Stack dump: stackPointer = 0

	000 SP-> 0


	 
	(Filename: C:/BuildAgent/work/d63dfc6385190b60/artifacts/StandalonePlayerGenerated/UnityEngineDebug.cpp Line: 49)

	kOS: Popping context 2
	 
	(Filename: C:/BuildAgent/work/d63dfc6385190b60/artifacts/StandalonePlayerGenerated/UnityEngineDebug.cpp Line: 49)

	kOS: Removed Context File                 Line:Col IP   opcode operand
	 
	(Filename: C:/BuildAgent/work/d63dfc6385190b60/artifacts/StandalonePlayerGenerated/UnityEngineDebug.cpp Line: 49)

	kOS: Deleting 0 pointers and restoring 1 pointers
	 
	(Filename: C:/BuildAgent/work/d63dfc6385190b60/artifacts/StandalonePlayerGenerated/UnityEngineDebug.cpp Line: 49)

	kOS: New current context File                 Line:Col IP   opcode operand
	 
	(Filename: C:/BuildAgent/work/d63dfc6385190b60/artifacts/StandalonePlayerGenerated/UnityEngineDebug.cpp Line: 49)

	kOS: FlightControl Binding
	 
	(Filename: C:/BuildAgent/work/d63dfc6385190b60/artifacts/StandalonePlayerGenerated/UnityEngineDebug.cpp Line: 49)

	kOS: FlightControl Bound

**Comment by Dunbaratu** [comment]
Okay, I have it trimmed down to this as the smallest example of the problem:
```
// A simple test of the for loop bug 
// from issue #694
//

// A simple list of mostly "A"s and one "B":
set myList to LIST().
myList:ADD("A").
myList:ADD("A").
myList:ADD("A").
myList:ADD("B").
myList:ADD("A").

// A for loop that will iterate until it hits a "B",
// then break:
for thing in MyList {
  print "thing=" + thing.
  if thing = "B" {
    break.
  }.
}.
print "done".
```

**Comment by Dunbaratu** [comment]
Okay here's an even smaller example of the problem:
```
set myList to LIST().
myList:add("A"). // have to have at least one thing for the FOR loop body to execute at all.
for thing in Mylist {
  break.
}.
```
Even that example triggers it (so it doesn't even have to be nested inside a further set of braces).

**Comment by Dunbaratu** [comment]
Okay so I think the problem is this:  In Compiler, When it does VisitForStatement, it tries to run some cleanup on the iterator by calling the enumertor:reset suffix after the loop is done, like so:
```
interpreter history     6:26  0057 push $ll-iterator 
interpreter history     6:26  0058 push reset 
interpreter history     6:26  0059 getmember
```
This is happening after the bottom of the loop, but I told the "break" statement that it should PopScope several levels all the way back out to outside this scope, even though it ends up jumping to this cleanup code rather than all the way out.  So the fix is to just tell it to set the break scope one more level deeper in (moving the PushBreakList() call one level of scope deeper in_.

That will fix the bug, but while I'm at it, this looks really pointless and could be optimized away - there's no need to reset the iterator after the loop is finished.  The iterator is going out of scope anyway.

**Comment by Dunbaratu** [comment]
github didn't auto-associate.  This issue should be closed when PR #695 is merged.
